kroxylicious-api/src/main/java/io/kroxylicious/proxy/service/Contributor.java

```diff
@@ -7,46 +7,39 @@
 
 import io.kroxylicious.proxy.config.BaseConfig;
 
+import edu.umd.cs.findbugs.annotations.NonNull;
+
 /**
  * Support loading an Instance of a service, optionally providing it with configuration obtained
  * from the Kroxylicious configuration file.
  *
- * @param <T> the service type
+ * @param <S> the service type
+ * @param <C> the context type
  */
-public interface Contributor<T, S extends Context> {
+public interface Contributor<S, B extends BaseConfig, C extends Context<B>> {
 
     /**
-     * Identifies if this Contributor instance can create instances for the provided short name.
-     * @param typeName The type name to test for.
-     * @return <code>true</code> if the <code>typeName</code> is supported by this contributor.
+     * Identifies the type name this contributor offers.
+     * @return typeName
      */
-    boolean contributes(String typeName);
+    @NonNull
+    String getTypeName();
 
-    /**
-     * @deprecated subsumed into {@link ConfigurationDefinition} and thus {@link Contributor#getConfigDefinition(String)}
-     *
-     * Gets the concrete type of the configuration required by this service instance.
-     * @param typeName service short name
-     *
-     * @return class of a concrete type, or null if this contributor does not offer this short name.
-     */
-    @Deprecated(forRemoval = true, since = "0.3.0")
-    Class<? extends BaseConfig> getConfigType(String typeName);
+    @NonNull
+    Class<B> getConfigType();
 
-    /**
-     * Defines the configuration requirements of this contributor for the given short name.
-     * @param typeName the name of the type.
-     * @return the ConfigurationDefinition for the short name
-     */
-    ConfigurationDefinition getConfigDefinition(String typeName);
+    @NonNull
+    default boolean requiresConfiguration() {
```

Comment:
The thing it's trying to signal is if the user declares their filter has a custom configuration type, this declares if the filter tolerates null being passed in the constructor.

Reply:
Right, yes, sorry. It's frustrating that Java's type system doesn't provide great support for this.

Reply:
Also more confusion since Keith added `If the service does not require configuration, return {@link Void} instead.` to `getConfigType`. Will do a pass over all the docs soon to clear it up.

Reply:
`Void.class` covers the cases where the filter does not accept configuration. Optional configuration is a different beast, it would still provide a custom type and we want a mechanism to validate the supplied configuration rather than just blowing up at construction time. 

Though I do agree that we need to handle problems constructing filters better.

Reply:
Maybe we're thinking about this wrong. I think we're agreed that there is value in being able to validate the config separately from actually instantiating anything (one advantage is it allows us to validate a config file at start up before there are any connections requiring filters). Perhaps we should reflect this in the `Contributor` interface?

```java
/**
 * Validate a configuration.
 * For any given `config`, this is guaranteed to be called before a call to {@link #getInstance(B)}.
 * @throws InvalidConfigurationException if the given config object is invalid
 */
void validateConfig(B config);
```

This would allow a `Contributor` to reject or accept a null `config` without having to model optionality it in either a `ContributorDefinition` or the `Contributor`, and without having to instantiate a service. It would also allow deeper semantic validation of configurations, which would otherwise be impossible to model in a `Definition`.

Reply:
Yup. I think that sort of model is the better long term option. 

I think I saw that as increasing the load on filter authors for the default case vs the rare case of optional config so went with the boolean which could be defaulted. However in hindsight I don't think that trade off was really valid.

Reply:
I think it might be useful to use the filter author quickstart as the means of making calls like that. i.e. write the code needed to demonstrate the contract we're providing, and the commentary that goes with it. I think it might illuminate the tradeoffs. 

Every time we expose the programmer to a new public type that's something they have to understand. And the complexity grows faster than linearly (they have to understand how that class relates to all the other classes they know about). In contrast, making them implement a `Contributor` with a couple of methods is conceptually lighter weight. I can imagine explaining that in the quickstart in a few sentences.

Reply:
maybe we can make it easy for the typical cases by implementing something like:

```
default void validateConfig(C config) {
        if (getConfigType() != Void.class && config == null) {
            throw new IllegalArgumentException();
        }
    }
```